src/main/scala/org/broadinstitute/dsde/firecloud/service/NihService.scala

```diff
@@ -127,6 +128,29 @@ class NihService(val samDao: SamDAO, val thurloeDao: ThurloeDAO, val googleDao:
     thurloeDao.saveKeyValues(userInfo, profilePropertyMap)
   }
 
+  private def unlinkNihAccount(userInfo: UserInfo): Future[Unit] = {
+    val nihKeys = Set("linkedNihUsername", "linkExpireTime")
+
+    Future.traverse(nihKeys) { nihKey =>
+      thurloeDao.deleteKeyValue(userInfo.id, nihKey, userInfo)
+    } map { results =>
+      if(!results.forall(_.isSuccess))
+        throw new FireCloudExceptionWithErrorReport(ErrorReport(StatusCodes.InternalServerError, "Unable to unlink NIH account"))
+    }
+  }
+
+  def unlinkNihAccountAndSyncSelf(userInfo: UserInfo): Future[Unit] = {
+    for {
+      _ <- unlinkNihAccount(userInfo)
+      _ <- ensureWhitelistGroupsExists()
+      _ <- Future.traverse(nihWhitelists) {
+        whitelist => removeUserFromNihWhitelistGroup(WorkbenchEmail(userInfo.userEmail), whitelist).recoverWith {
```

Comment:
Similar to above, is it a concern if deleting the keys in Thurloe succeeds but removing from the groups fails?

Reply:
Deleting the keys but not removing from the group is okay imo, the process that syncs NIH groups will realize it within 6 hours and correct it.